sunpy/net/tests/test_scraper.py

```diff
@@ -225,3 +227,45 @@ def test_yearly_overlap():
     # Should return a single file for 2013
     trange = TimeRange("2013-01-02", "2013-01-03")
     assert len(scraper.filelist(trange)) == 1
+    
+#this fixture is used to test the case when the server returns http errors like 400, 403, 429, 504
+#the first element of the tuple is the error code and the second element is the number of times the urlopen function is called
+#the urlopen function is called multiple times because the scraper retries the request so the test asserts the max retry limit
+@pytest.fixture(params=[(400,1) , (403,1) , (429,6) , (504,6)])
+def endpoint(request):
+    return request.param
+
+def test_http_errors_with_enqueue_limit(endpoint):
+    with patch('sunpy.net.scraper.urlopen') as mocked_urlopen:
+        mocked_urlopen.side_effect = HTTPError('http://example.com', endpoint[0], '', {}, None)
+        time = TimeRange('2012/3/4', '2012/3/4 02:00')
+        pattern = "http://proba2.oma.be/lyra/data/bsd/{{year:4d}}/{{month:2d}}/{{day:2d}}/{{}}_lev{{Level:1d}}_std.fits"
+        scraper = Scraper(pattern)
+        with pytest.raises(HTTPError) as excinfo:
+            scraper._httpfilelist(time)
+        assert excinfo.value.code == endpoint[0]
+        assert mocked_urlopen.call_count == endpoint[1]
+
+def test_connection_error():
+    with patch('sunpy.net.scraper.urlopen') as mocked_urlopen:
+        mocked_urlopen.side_effect = URLError('connection error')
+        time = TimeRange('2012/3/4', '2012/3/4 02:00')
+        pattern = "http://proba2.oma.be/lyra/data/bsd/{{year:4d}}/{{month:2d}}/{{day:2d}}/{{}}_lev{{Level:1d}}_std.fits"
+        scraper = Scraper(pattern)
+        with pytest.raises(URLError, match = 'connection error'):
+            scraper._httpfilelist(time)
+            
+def test_http_404_error_debug_message(caplog):
+    log = logging.getLogger('sunpy')
+    log.setLevel(logging.DEBUG)
+    def patch_range(self, range):
+        return ['http://test.com/']
+    with patch('sunpy.net.scraper.urlopen') as mocked_urlopen:
+        with patch.object(Scraper, 'range', patch_range):
+            mocked_urlopen.side_effect = HTTPError('http://example.com', 404, '', {}, None)
+            time = TimeRange('2012/3/4', '2012/3/4 02:00')
+            pattern = "http://proba2.oma.be/lyra/data/bsd/{{year:4d}}/{{month:2d}}/{{day:2d}}/{{}}_lev{{Level:1d}}_std.fits"
+            scraper = Scraper(pattern)
+            scraper._httpfilelist(time)
+            assert "Directory http://test.com/ not found." in caplog.text
+    log.setLevel(logging.NOTSET)
```

Comment:
Actually can you get the original logging level at the start of the test and set it to that value? We should make sure we don't change it.

Reply:
The test should try to not change any setting that might affect another test.

So we should reset the level to the original value at the end.